server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java

```diff
@@ -0,0 +1,1350 @@
+   * @return the number of rows persisted
+   */
+  private int persistHydrant(FireHydrant indexToPersist, SegmentIdWithShardSpec identifier)
+  {
+    synchronized (indexToPersist) {
+      if (indexToPersist.hasSwapped()) {
+        log.info(
+            "Segment[%s] hydrant[%s] already swapped. Ignoring request to persist.",
+            identifier,
+            indexToPersist
+        );
+        return 0;
+      }
+
+      log.debug("Segment[%s], persisting Hydrant[%s]", identifier, indexToPersist);
+
+      try {
+        final long startTime = System.nanoTime();
+        int numRows = indexToPersist.getIndex().size();
+
+        // since the sink may have been persisted before it may have lost its
+        // hydrant count, we remember that value in the sinks metadata so we have
+        // to pull it from there....
+        SinkMetadata sm = sinksMetadata.get(identifier);
+        final File persistDir = createPersistDirIfNeeded(identifier);
+        indexMerger.persist(
+            indexToPersist.getIndex(),
+            identifier.getInterval(),
+            new File(persistDir, String.valueOf(sm.getNumHydrants())),
+            tuningConfig.getIndexSpecForIntermediatePersists(),
+            tuningConfig.getSegmentWriteOutMediumFactory()
+        );
+
+        log.info(
+            "Persisted in-memory data for segment[%s] spill[%s] to disk in [%,d] ms (%,d rows).",
+            indexToPersist.getSegmentId(),
+            indexToPersist.getCount(),
+            (System.nanoTime() - startTime) / 1000000,
+            numRows
+        );
+
+        indexToPersist.swapSegment(null);
+        // remember hydrant count:
+        sinksMetadata.get(identifier).addHydrants(1);
+
+        return numRows;
+      }
+      catch (IOException e) {
+        log.makeAlert("Incremental persist failed")
+           .addData("segment", identifier.toString())
+           .addData("dataSource", schema.getDataSource())
+           .addData("count", indexToPersist.getCount())
+           .emit();
+
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private void removeDirectory(final File target)
+  {
+    if (target.exists()) {
+      try {
+        FileUtils.deleteDirectory(target);
+        log.info("Removed directory [%s]", target);
+      }
+      catch (Exception e) {
+        log.makeAlert(e, "Failed to remove directory[%s]", schema.getDataSource())
+           .addData("file", target)
+           .emit();
+      }
+    }
+  }
+
+  private int calculateMemoryUsedByHydrants(FireHydrant hydrant)
+  {
+    if (skipBytesInMemoryOverheadCheck) {
+      return 0;
+    }
+    // These calculations are approximated from actual heap dumps.
+    // Memory footprint includes count integer in FireHydrant, shorts in ReferenceCountingSegment,
+    // Objects in SimpleQueryableIndex (such as SmooshedFileMapper, each ColumnHolder in column map, etc.)
+    int total;
+    total = Integer.BYTES + (4 * Short.BYTES) + ROUGH_OVERHEAD_PER_HYDRANT;
```

Comment:
Is this calculation still accurate?

Reply:
Yeah, it is good enough. Before it used to track memory used by memory mapped references which are no longer there.

Reply:
nit: seems like a lot of lines for a constant. if i understand correctly there is only 1 hydrant per sink, would it just make sense to factor this into the sink calculation?

Also the comment doesn't seem relevant and should be fixed or removed